File: server/languages/java/src/main/java/LanguageModuleImp.java
import java.util.Arrays;
import io.grpc.stub.StreamObserver;
import com.google.protobuf.ByteString;

public class LanguageModuleImp extends LanguageModuleGrpc.LanguageModuleImplBase
{
	@Override
	public void getCapabilities(Common.Empty request, StreamObserver<Common.LanguageCapabilities> responseObserver)
	{
		Common.LanguageCapabilities response = Common.LanguageCapabilities.newBuilder()
			.setLanguage("java")
			.setCapabilities(Common.Capabilities.GENERATE_ASTS_VALUE | Common.Capabilities.IO_MATCHING_VALUE | Common.Capabilities.UNIT_TESTING_VALUE)
			.build();
		responseObserver.onNext(response);
		responseObserver.onCompleted();
	}
	
	@Override
	public void getSandboxDetails(Common.Empty request, StreamObserver<Language.SandboxDetails> responseObserver)
	{
		Language.SandboxDetails response = Language.SandboxDetails.newBuilder()
			.setImage("adamrehn/language-toolbox-sandbox-java")
			.addAllCommand(Arrays.asList(new String[]{
				"bash", "-c", "cat > /tmp/temp.jar && java -jar /tmp/temp.jar"
			}))
			.build();
		responseObserver.onNext(response);
		responseObserver.onCompleted();
	}
	
	public void generateAst(Common.GenerateAstRequest request, StreamObserver<Common.GenerateAstResponse> responseObserver)
	{
		Common.GenerateAstResponse response = null;
		
		try
		{
			String ast = AstTransformer.generateAst(request.getSource());
			response = Common.GenerateAstResponse.newBuilder()
				.setError("")
				.setAst(ast)
				.build();
		}
		catch (Exception e)
		{
			response = Common.GenerateAstResponse.newBuilder()
				.setError(e.toString())
				.setAst("")
				.build();
		}
		
		responseObserver.onNext(response);
		responseObserver.onCompleted();
	}
	
	@Override
	public void codegenIOCapture(Language.CodegenIOCaptureRequest request, StreamObserver<Language.CodegenIOCaptureResponse> responseObserver)
	{
		Language.CodegenIOCaptureResponse response = null;
		
		try
		{
			byte[] codegenResult = IOCaptureCodegen.performCodegen(request.getSource(), request.getInvocation(), request.getStdin());
			response = Language.CodegenIOCaptureResponse.newBuilder()
				.setError("")
				.setData(ByteString.copyFrom(codegenResult))
				.build();
		}
		catch (Exception e)
		{
			response = Language.CodegenIOCaptureResponse.newBuilder()
				.setError(e.toString())
				.setData(ByteString.EMPTY)
				.build();
		}
		
		responseObserver.onNext(response);
		responseObserver.onCompleted();
	}
	
	@Override
	public void codegenUnitTests(Language.CodegenUnitTestsRequest request, StreamObserver<Language.CodegenUnitTestsResponse> responseObserver)
	{
		Language.CodegenUnitTestsResponse response = null;
		
		try
		{
			byte[] codegenResult = UnitTestCodegen.performCodegen(request.getSource(), request.getTestsList());
			response = Language.CodegenUnitTestsResponse.newBuilder()
				.setError("")
				.setData(ByteString.copyFrom(codegenResult))
				.build();
			
		}
		catch (Exception e)
		{
			response = Language.CodegenUnitTestsResponse.newBuilder()
				.setError(e.toString())
				.setData(ByteString.EMPTY)
				.build();
		}
		
		responseObserver.onNext(response);
		responseObserver.onCompleted();
	}
}

File: server/languages/cxx/source/LanguageModuleImp.cpp
#include "LanguageModuleImp.h"
#include "AstGenerator.h"
#include "IOCaptureCodegen.h"
#include "UnitTestCodegen.h"

grpc::Status LanguageModuleImp::GetCapabilities(ServerContext* context, const Empty* request, LanguageCapabilities* response)
{
	response->set_language("cxx");
	response->set_capabilities(Capabilities::GENERATE_ASTS | Capabilities::IO_MATCHING | Capabilities::UNIT_TESTING);
	return grpc::Status::OK;
}

grpc::Status LanguageModuleImp::GetSandboxDetails(ServerContext* context, const Empty* request, SandboxDetails* response)
{
	response->set_image("adamrehn/language-toolbox-sandbox-cxx");
	response->add_command("bash");
	response->add_command("-c");
	response->add_command("cat > /tmp/a.out && chmod a+x /tmp/a.out && /tmp/a.out");
	return grpc::Status::OK;
}

grpc::Status LanguageModuleImp::GenerateAst(ServerContext* context, const GenerateAstRequest* request, GenerateAstResponse* response)
{
	//Attempt to generate the AST
	try
	{
		response->set_error("");
		response->set_ast(AstGenerator::generateAst(request->source()));
	}
	catch (std::runtime_error& e)
	{
		response->set_error(e.what());
		response->set_ast("");
	}
	
	return grpc::Status::OK;
}

grpc::Status LanguageModuleImp::CodegenIOCapture(ServerContext* context, const CodegenIOCaptureRequest* request, CodegenIOCaptureResponse* response)
{
	//Attempt to perform codegen
	try
	{
		response->set_error("");
		response->set_data(IOCaptureCodegen::performCodegen(&this->codegen, request->source(), request->invocation(), request->stdin()));
	}
	catch (std::runtime_error& e)
	{
		response->set_error(e.what());
		response->set_data("");
	}
	
	return grpc::Status::OK;
}

grpc::Status LanguageModuleImp::CodegenUnitTests(ServerContext* context, const CodegenUnitTestsRequest* request, CodegenUnitTestsResponse* response)
{
	//Retrieve the list of unit tests
	vector<UnitTest> tests;
	for (int i = 0; i < request->tests_size(); ++i) {
		tests.push_back(request->tests(i));
	}
	
	//Attempt to perform codegen
	try
	{
		response->set_error("");
		response->set_data(UnitTestCodegen::performCodegen(&this->codegen, request->source(), tests));
	}
	catch (std::runtime_error& e)
	{
		response->set_error(e.what());
		response->set_data("");
	}
	
	return grpc::Status::OK;
}
